### PR TITLE
`Integrated code lifecycle`: Configure checkout path and timeout for programming exercises

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Second PR of this feature proposal: #8887

We want the instructors to be able to customize the build timeout and the checkout paths

### Quick testing (On test servers)
Prerequisites:
- 1 Instructor

## Checkout Paths testing

1. Go to programming exercise creation
2. Change the checkout paths:
    - Adding nested paths (e.g. assignment_1/assignment_2)  may lead to a build failure. This is expected as it is up for the instructor to adjust the test code/script.
    - For testing purposes, just change the name without nesting (for example change 'assignment' to 'exerciseCode')
3. Click on `Customize Build Script`, If the build script uses the default checkout paths (assignment/solution/tests) make sure to change them accordingly
4. Create exercise and make sure that builds don't fail. However, we expect failures in these programming languages: Haskell, C.
5. Repeat previous steps for different programming languages and project types, please write the ones you tested in your comment.

## Timeout testing

1. Edit the timeout, and then run a build for that exercise. The build duration may exceed the timeout as there are pre-processing steps, the cancellation steps take a few seconds.
2. Make sure that other programming exercises (those that don't have a checkout path/timeout set) work normally.

## Testing Locally (Optional)
Prerequisites:
- 1 Instructor
- Go to BuildJobContainerService.java, comment out line 104 and uncomment line 105
```
//                .withCmd("sh", "-c", "while [ ! -f " + workingDirectory + "/stop_container.txt ]; do sleep 0.5; done")
                 .withCmd("tail", "-f", "/dev/null") // Activate for debugging purposes instead of the above command to get a running container that you can peek into using
``` 
- Go to BuildAgentConfiguration.java, change line 92 to `.withAutoRemove(false);`

1. Create programming exercise
2. Change checkout paths. 
3. In the container's filesystem, check that the exercise was cloned in the right place. 
  - This can be checked in docker desktop (click on the container, then click on files, you navigate the file system there)
  - If you don't have docker desktop, run `docker exec -it <mycontainer> sh` Navigate using the cli

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| programming-exercise-update.component.ts | 87.89% |  ✅ |
| programming-exercise-build-configuration.component.ts | 100% | ✅ |
| programming-exercise-custom-aeolus-build-plan.component.ts | 95.34% | ✅ |
| programming-exercise-custom-build-plan.component.ts | 98.55% | ✅ |
| programming-exercise-information.component.ts | 94.11% | ✅  |
| programming-exercise-language.component.ts | 88.37% | ✅ |
| programming-exercise-edit-checkout-directories.component.ts | 98% | ✅  |
| programming-exercise-repository-and-build-plan-details.component.ts | 87.5% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| BuildJobContainerService.java | 81% | ✅ |
| BuildJobExecutionService.java | 75% | ✅ |
| BuildJobManagementService.java | 65% | ✅ |
| BuildScriptProviderService.java | 89% | ❌  |
| ProgrammingExerciseRepositoryService.java | 82% | ✅ |
| ProgrammingExerciseService.java | 87% | ✅ |
| LocalCIBuildConfigurationService.java | 76% | ✅ |
| LocalCITriggerService.java | 88% | ✅  |

### Screenshots

image

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced the build job container setup with customizable checkout paths for assignment, test, and solution repositories.
	- Improved script handling with dynamic placeholder replacements based on the CI environment.
	- Enhanced directory path handling for programming exercises, accommodating various configurations.
	- Introduced stricter validation for editing programming exercises, including timeout and checkout path checks.
	- Added localization support for Docker configurations and build parameters in English and German.
	- Expanded the functionality of the build configuration component to manage timeout settings.
	- Introduced a new component for editing checkout directories associated with programming exercises.
	- Added user interface enhancements for configuring Docker images and timeout settings in programming exercises.

- **Bug Fixes**
	- Added tests to validate the functionality of the new repository checkout path features.
	- Implemented additional validation checks for timeout and checkout paths in the programming exercise management.

- **Documentation**
	- Updated documentation to reflect changes in repository management and configuration options.

- **Refactor**
	- Improved code readability and modularity through the introduction of utility methods for path manipulation.
	- Reorganized components within the programming exercise update module to emphasize build configurations.

### CODE REVIEW

File: src/main/resources/templates/python/test/structural/structural_test.py
Comment (SimonEntholzer) [inline src/main/resources/templates/python/test/structural/structural_test.py:5]: I don't fully get how this works here, but I guess I missed the part where the files (in this case .py) are transformed to a correct file. Where does this happen?

Comment (BBesrour) [reply src/main/resources/templates/python/test/structural/structural_test.py:5]: This will be replace during the exercise creation. Python has a special case where the path (e.g. assignment/code) has to be change to (assignment.code). This happens in [this file](https://github.com/ls1intum/Artemis/pull/9217/files#diff-20411055de836e298f72ea8eaa62b094edeeae51b7073d159b9ef9572d819b9f)

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Seems to work fine on Ts4, a few suggestions though:

Probably it makes sense to move the button up, when the create auxiliary repository dialog is open:
image

There seems to be something broken when I create an exercise with auxiliary repository (I created it with haskell)
When saving, no aux repos show up any more. When going to edit mode, I also can't save any more, as the urls changed back to the default.
image

**Comment by BBesrour** [comment]
@undernagruzez yes, since we changed the templates

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
I think there might be still an issue with the C exercise, at least the build fails:
image
image
I get more or less the same issue with the VHDL exercise. 
Build for Ocaml fails too, seems like it can't find some srcipt as well.

<details>

```

Build Error:

2024-09-15 12:00:08
    ~~~~~~~~~~~~~~~~~~~~ Start Build Job 22491726394408631 ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:08
    ~~~~~~~~~~~~~~~~~~~~ Inspecting docker image ls1tum/artemis-ocaml-docker:v1 ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:08
    ~~~~~~~~~~~~~~~~~~~~ Inspecting docker image ls1tum/artemis-ocaml-docker:v1 again with a lock due to error Status 404: {"message":"No such image: ls1tum/artemis-ocaml-docker:v1"} ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:08
    ~~~~~~~~~~~~~~~~~~~~ Pulling docker image ls1tum/artemis-ocaml-docker:v1 with a lock after error Status 404: {"message":"No such image: ls1tum/artemis-ocaml-docker:v1"} ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:37
    ~~~~~~~~~~~~~~~~~~~~ Pull image complete ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:37
    ~~~~~~~~~~~~~~~~~~~~ Pulling docker image ls1tum/artemis-ocaml-docker:v1 done after 28.55sec ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:40
    ~~~~~~~~~~~~~~~~~~~~ Started container local-ci-22491726394408631 for build job 22491726394408631 ~~~~~~~~~~~~~~~~~~~~ 
    ~~~~~~~~~~~~~~~~~~~~ Populating build job container with repositories and build script ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:40
    ~~~~~~~~~~~~~~~~~~~~ Executing Build Script for Build job 22491726394408631 ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:40
    ⚙️ executing build_and_test_the_code 
    /var/tmp/script.sh: line 8: cd: tests: No such file or directory 
2024-09-15 12:00:40
    ⚙️ executing final_aeolus_post_action ⚙️ executing junit 
    ~~~~~~~~~~~~~~~~~~~~ Finished Executing Build Script for Build job 22491726394408631 ~~~~~~~~~~~~~~~~~~~~ 
2024-09-15 12:00:41
    Building and testing submission for repository ga84tactestcoursetestnewnamingfeaturea7-solution and commit hash 70ae5875c1ff6dfd337fb42f00e8db2fb5f854ec took 1.33sec 

```


</details>

The haskell exercise build fails too, but there I'm not sure if that's related to this PR.

**Comment by BBesrour** [comment]
@SimonEntholzer Thnx for the review, C exercises fail because there is a built-in library in the docker image that expects the student's code in a specific path, we can't change that unfortunately. For haskell, builds also fail on develop. For Ocaml, unfortunately, it's difficult to dynamically adjust the build script with current implementation, so Instructors will have to adjust it manually if they change the checkout paths.

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java
Comment (coolchock) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:294]: same here

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java
Comment (coolchock) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:307]: also here

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseService.java
Comment (coolchock) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseService.java:445]: NIT: I think it would be helpful to add a short comment here. Initially, it wasn't clear to me why a null checkoutPath would be considered valid.

File: src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingExerciseBuildConfig.java
Comment (coolchock) [inline src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingExerciseBuildConfig.java:52]: NIT: I think we could also update the column name, but I don't have a strong opinion on that

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (coolchock) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:1121]: should we check for blank strings there too?

File: src/main/webapp/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-custom-build-plan.component.ts
Comment (coolchock) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-custom-build-plan.component.ts:66]: blank check?

File: src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-edit-checkout-directories/programming-exercise-edit-checkout-directories.component.ts
Comment (coolchock) [inline src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-edit-checkout-directories/programming-exercise-edit-checkout-directories.component.ts:48]: blank check?

File: src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-edit-checkout-directories/programming-exercise-edit-checkout-directories.component.ts
Comment (coolchock) [inline src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-edit-checkout-directories/programming-exercise-edit-checkout-directories.component.ts:105]: blank check?

File: src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts
Comment (coolchock) [inline src/main/webapp/app/exercises/programming/shared/build-details/programming-exercise-repository-and-build-plan-details.component.ts:131]: blank checks?

**Review by coolchock — CHANGES_REQUESTED** [review]
code LGTM, I only left a few remarks

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:290]: Also use the `StringUtils`, since they also do the null-check for you

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:446]: Can we use the actual Path class. The handling with slashes is hacky, since Windows uses \ instead of /

File: src/main/java/de/tum/cit/aet/artemis/programming/service/BuildScriptProviderService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/BuildScriptProviderService.java:208]: Please use `StringUtils` for the implicit null-check

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseRepositoryService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseRepositoryService.java:670]: `StringUtils`

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseService.java
Comment (JohannesStoehr) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseService.java:445]: Depending on your preferences we could also use Optional or @Nullable to make it clearer that they do not have to be set

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:446]: This paths will be used inside a docker container so it'll be UNIX paths. At first I implemented using paths but it was unnecessarily long and complicated

File: src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingExerciseBuildConfig.java
Comment (BBesrour) [reply src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingExerciseBuildConfig.java:52]: Change the entry in the database would make it harder to test on TS since deployed this PR would break the TS for other PRs

**Review by coolchock — DISMISSED** [review]
thanks for the changes! code LGTM now 👍

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by coolchock — DISMISSED** [review]
re-approve after JS template was added

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code

**Review by SimonEntholzer — APPROVED** [review]
Works on Ts2, Ocaml works now by adjusting the build script :+1:

**Review by edkaya — DISMISSED** [review]
Tested on ts2, works as expected

**Review by SimonEntholzer — DISMISSED** [review]
retested + reapprove

**Review by JohannesStoehr — DISMISSED** [review]
Reapprove code after merge

**Review by coolchock — APPROVED** [review]
re-approve after merge conflict

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Review by edkaya — DISMISSED** [review]
reapprove

**Review by SimonEntholzer — APPROVED** [review]
reapprove

**Review by edkaya — APPROVED** [review]
reapprove after merge conflict

**Review by SimonEntholzer — APPROVED** [review]
reapprove after merge conflicts, and short retest on Ts2

**Pull request merged**
